pubsublite-spark-sql-streaming/src/main/java/com/google/cloud/pubsublite/spark/PslDataSourceOptions.java

```diff
@@ -0,0 +1,121 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.pubsublite.spark;
+
+import static com.google.cloud.pubsublite.internal.ServiceClients.addDefaultSettings;
+
+import com.google.auto.value.AutoValue;
+import com.google.cloud.pubsublite.SubscriptionPath;
+import com.google.cloud.pubsublite.cloudpubsub.FlowControlSettings;
+import com.google.cloud.pubsublite.v1.AdminServiceClient;
+import com.google.cloud.pubsublite.v1.AdminServiceSettings;
+import com.google.cloud.pubsublite.v1.CursorServiceClient;
+import com.google.cloud.pubsublite.v1.CursorServiceSettings;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Optional;
+import javax.annotation.Nullable;
+import org.apache.spark.sql.sources.v2.DataSourceOptions;
+
+@AutoValue
+public abstract class PslDataSourceOptions implements Serializable {
+  private static final long serialVersionUID = 2680059304693561607L;
+
+  @Nullable
+  public abstract String credentialsKey();
+
+  public abstract SubscriptionPath subscriptionPath();
+
+  @Nullable
+  public abstract FlowControlSettings flowControlSettings();
+
+  public abstract long maxBatchOffsetRange();
+
+  public static Builder builder() {
+    return new AutoValue_PslDataSourceOptions.Builder()
+        .credentialsKey(null)
+        // TODO(jiangmichael): Revisit this later about if we need to expose this as a user
+        // configurable option. Ideally we should expose bytes range/# msgs range not
+        // offsets range since PSL doesn't guarantee offset = msg.
+        .maxBatchOffsetRange(Constants.DEFAULT_BATCH_OFFSET_RANGE);
+  }
+
+  public static PslDataSourceOptions fromSparkDataSourceOptions(DataSourceOptions options) {
+    if (!options.get(Constants.SUBSCRIPTION_CONFIG_KEY).isPresent()) {
+      throw new IllegalArgumentException(Constants.SUBSCRIPTION_CONFIG_KEY + " is required.");
+    }
+
+    Builder builder = builder();
+    Optional<String> value;
+    if ((value = options.get(Constants.CREDENTIALS_KEY_CONFIG_KEY)).isPresent()) {
+      builder.credentialsKey(value.get());
+    }
+    return builder
+        .subscriptionPath(
+            SubscriptionPath.parse(options.get(Constants.SUBSCRIPTION_CONFIG_KEY).get()))
+        .flowControlSettings(
+            FlowControlSettings.builder()
+                .setMessagesOutstanding(
+                    options.getLong(
+                        Constants.MESSAGES_OUTSTANDING_CONFIG_KEY,
+                        Constants.DEFAULT_MESSAGES_OUTSTANDING))
+                .setBytesOutstanding(
+                    options.getLong(
+                        Constants.BYTES_OUTSTANDING_CONFIG_KEY,
+                        Constants.DEFAULT_BYTES_OUTSTANDING))
+                .build())
+        .build();
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+
+    public abstract Builder credentialsKey(String credentialsKey);
```

Comment:
nit. these methods should be called "set<thing>"

Reply:
Done